### DISCUSSION TITLE
The Inline Python Code for Quarto is not working in VS Code but the Inline R Code is!

### DISCUSSION THREAD

**Discussion by victorvanpelt** [quarto-dev/quarto-cli#9731] [Q&A] [answered]
### Description

Hello,

I am currently incredibly confused why my inline python code is not working, but my inline R code is working.

Here is my code:  

![code](https://github.com/quarto-dev/quarto-cli/assets/13329496/7c13ab56-97db-4858-85ff-b43835ad6fdb)

This produces the following:  

![test](https://github.com/quarto-dev/quarto-cli/assets/13329496/579dc39c-8778-47fd-951b-6a2516f81335)

Does anyone know why my python inline code isn't working? 

I have already checked whether Python executes. It does, but only within the python code block.

Thanks in advance!

**Comment by mcanouil** [comment]
The Python inline is for Jupyter (https://quarto.org/docs/computations/inline-code.html).

I don't think knitr/reticulate even supports Python inline or any knitr engine/language other than R.
Cc @cderv

**Reply by victorvanpelt** [reply]
Thanks for responding. I'm new to Quarto! So, python inline code only works for Jupiter. R does work in the other engines? 

I'm a little confused because the link you shared (which I also read) basically says python SHOULD work across all three engines:
"Quarto provides an inline code syntax that works across all three engines (Jupyter, Knitr and OJS)." I tried changing the engine (i.e., jupyter:python3), but that doesn't do much.

**Reply by mcanouil** [reply]
The syntax works across all engines. Languages support needs to come from the engine themselves.
Quarto can't really know what language your engine supports or not to do its magic.

- `knitr` engine has support for third party R packages which brings other language support all through R, via the R package `reticulate` for Python and `JuliaCall` for Julia.
- Jupyter engine depends on the kernels (there are quite a bunch out-there). With the Python kernel, you can only execute Python code. If you want to execute R code via Python Jupyter kernel, then you need a Python module such as `rpy2`.
- The Julia kernel for Jupyter or the new Julia native engine (1.5 pre-release) would also needs third party to execute "foreign" code.

**Reply by cscheid** [reply]
Hm. @mcanouil I think this is a bug in Quarto. There's an interaction with Reticulate, as you've noted, but we should be changing `{python} x` to `python x` before knitr/reticulate sees the code, and then the inline expressions should work.

**Reply by mcanouil** [reply]
I was waiting for the expert I cc'ed in my initial reply.

**Reply by cderv** [reply]
> "Quarto provides an inline code syntax that works across all three engines (Jupyter, Knitr and OJS)."

What this sentence says is that the syntax `` `{language} code` is the same across engine (Jupyter, Knitr, OJS). It does not details how this works in details and the multi language document case. 

`{r} x` would no work in a `jupyter: python3` engine document, nor the R cells . Python cells and R cells are possibly together in same document using the knitr engine, thanks to **reticulate** R package integration. 

For Jupyter this would require custom kernel that allows multi language, though I don't know how inline code would behave in such context.

**Reply by cderv** [reply]
> we should be changing {python} x to python x before knitr/reticulate sees the code, and then the inline expressions should work.

@cscheid this does not work that way. `` `python x` `` is not valid syntax. When reticulate is used (which is the default), `` `r py$x` `` would work. So we could be changing `` `{python} x` `` to this in Quarto when inside knitr engine context.

**Reply by mcanouil** [reply]
> So we could be changing `{python} x` to this in Quarto when inside knitr engine context.

You mean `` `{python} x` `` to `` `r py$x` `` (internally)?

**Reply by cderv** [reply]
yes - something like that, when we are sure that **reticulate** is available and loaded.

**Reply by cscheid** [reply]
@cderv Thanks for the correction! Is there a syntax that works for a more general expression? Say, `{python} x * foo(3 + 4, bar = y + "blah")`

**Reply by cderv** [reply]
For this, it would mean asking **reticulate** to execute the python code from R : https://rstudio.github.io/reticulate/articles/calling_python.html

There are several ways, one of them is to execute from a string - really dummy example

````markdown
---
title: "A title"
---


```{r}
library(reticulate)
```

```{python}
x = 5
y = "boo"
## foo is a function that takes x and y and concatenating both returning a string
def foo(x, bar):
  return str(x) + "**" + str(bar) + "**" + " "
```

`r py_run_string('zz = x * foo(3 + 4, bar = y + "blah")'); py$zz`
````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/9384f044-e36e-440d-9a19-d64753e99e81)

Other way is to run from a file. 

So we could do something more complex, like writing to temp file the code, running it with reticulate and loading the object. 

I wonder if this should be something from **reticulate** itself instead of Quarto, through a more generic support in **knitr**
* https://github.com/yihui/knitr/issues/2026

**Reply by cscheid** [reply]
Let's discuss this on Thursday.

**Comment by cderv** [comment, marked as answer]
> I don't think knitr/reticulate even supports Python inline or any knitr engine/language other than R.

Let me be more precise on this affirmation. 

**knitr** supports any language that has a knitr's engine defined. Built-in or coming from other R packages. More on this in **knitr** docs
 - https://bookdown.org/yihui/rmarkdown/language-engines.html
 - https://bookdown.org/yihui/rmarkdown-cookbook/other-languages.html
Even a custom engine can be defined (https://bookdown.org/yihui/rmarkdown-cookbook/custom-engine.html)

`python` engine for **knitr** is from **reticulate** R package: https://rstudio.github.io/reticulate/reference/eng_python.html

Regarding inline execution, it is a **knitr** specific feature to the R engine - so only `` `r x` `` works (or `` `{r} x` `` in quarto  - https://quarto.org/docs/computations/inline-code.html#syntax-compatibility) - at least for now https://github.com/yihui/knitr/issues/2026

However, some packages like **reticulate** offers interoperatibility, and so Python values can be accessed from R. This means you just have to leverage the R inline syntax. See https://rstudio.github.io/reticulate/articles/r_markdown.html

````markdown
---
title: test
---

```{r}
library(reticulate)
```


```{python}
a = 5
```

The variable is `r py$a`

```{r}
b = 5
```

The variable is `r b`

````

I hope this helps understand how it works.

**Reply by mcanouil** [reply]
I knew, there was a trick! Thanks Christophe!

**Reply by victorvanpelt** [reply]
Thanks a lot for explaining!

**Reply by MattFloodComReg** [reply]
Hi Guys,
Thanks for the detailed explanation above @cderv :)
I'm experiencing the same problem as the original poster, but the above solution works using `r py$...` syntax to execute the python.
Perhaps the [docs](https://quarto.org/docs/computations/inline-code.html) should be updated to reflect this...?
